### ISSUE TITLE
CLI Revamp

### ISSUE THREAD

**Issue by rcdailey** [recyclarr/recyclarr#142]
## How To Read

Command line is described as follows:

- Portions shown between square brackets (e.g. `[--argument]`) are *optional*.
- Portions shown between angle brackets (e.g. `<path>`) are *required*.
- Portions shown without any brackets are required only by context; they form the basis of the
  command being documented (e.g. `recyclarr sync`).
- The pipe character (`|`) denotes a short & long form of an argument.

## New CLI

The CLI should be revamped to be more consistent and logically structured. Right now, the Docker CLI
is a motivation for the design.

Proposal for new CLI for the 3.0 release is documented in the sections below.

### Subcommands

- `sync`
- `config`
- `migrate`
- `list`

### Argument Changes

Arguments will now belong to specific stems of the command chain. For example, instead of `recyclarr
radarr --app-data my-dir`, you'll do `recyclarr --app-data my-dir radarr` because the `--app-data`
argument belongs to `recyclarr`, not `radarr`.

### Recyclarr (Main Command)

```sh
recyclarr [-d|--debug] [--app-data <path>]
```

- `--debug`: Debug logs are shown in console output.
- `--app-data <path>`: Path to the application data directory.

### Sync

```sh
recyclarr sync [service] [-n|--names <name> [name2 [...]]] [-c|--configs <yaml>] [-p|--preview]
```

Sync the guide to services.

- `[service]`: One of either `sonarr` or `radarr`. If not specified, defaults to *all* services.
- `--names`: One or more instance names to sync.
- `--preview`: Perform a dry run: preview the results without syncing.
- `--configs`: One or more YAML configuration files to load & use.

### List

```sh
recyclarr list custom-formats <service>
recyclarr list release-profiles [--terms <rp_trash_id>]
recyclarr list qualities <service>
```

List information from the guide.

- `<service>`: One of either `sonarr` or `radarr`. If a specific list type does not take a service
  argument, then that service is specific to only one kind of service and that is implied on the
  CLI. For example, Release Profiles are always assumed to be for Sonarr.
- `--terms`: List the identifiable terms in a Release Profile by its Trash ID.

### Config

```sh
recyclarr config create [-p|--path <path>]
```

Create a starter configuration file.

- `--path`: Path to where the configuration file should be created.

### Migrate

```sh
recyclarr migrate
```

Perform migration steps, if needed.

## Backward Compatibility

How old commands map to the new CLI and will continue to function. This list isn't comprehensive,
rather just gives some examples. It is expected that every current CLI command continues to work, but they will be deprecated and a warning will be printed in the logs when they are used.

### Sync Radarr

```sh
# Old Way
recyclarr radarr

# New Way
recyclarr sync radarr
```

### Sync Sonarr

```sh
# Old Way
recyclarr sonarr

# New Way
recyclarr sync sonarr
```

### Sync Both Radarr & Sonarr

```sh
# Old Way
recyclarr radarr && recyclarr sonarr

# New Way
recyclarr sync
```

### Create Starter Config

```sh
# Old Way
recyclarr create-config

# New Way
recyclarr config create
```

### List Radarr CFs

```sh
# Old Way
recyclarr radarr --list-custom-formats

# New Way
recyclarr list custom-formats radarr
```

### List Sonarr RPs With Debug Logs

```sh
# Old Way
recyclarr sonarr --debug --list-release-profiles

# New Way
recyclarr --debug list release-profiles
```

**Comment by rcdailey** [comment]
> 1. * a possible typo/wrong command in the old/new comparison section - should it be:
>      Old Way
>      recyclarr radarr && **recyclarr** sonarr

You are right; I fixed it. Thanks!

> 2. * Re: some related discussion elsewhere that 3.0 might support defining a location, and then all .yml files in that location would be processed as indicated by the preceding commands. Is that covered in this, or is that a future thing?

I don't understand. Are you referring to [the `configs` directory](https://recyclarr.dev/wiki/file-structure#directory-configs)?

**Comment by yammes08** [comment]
> I don't understand. Are you referring to [the `configs` directory](https://recyclarr.dev/wiki/file-structure/#config-directory)?

Yep! And that's covered so all looks good to me.

**Comment by snoopy82481** [comment]
Do you want the wiki to be formatted just like that?  I can create a page like that and have a pr for it, then when the day comes you can commit it.

**Comment by rcdailey** [comment]
I'll probably wait until the code is written and closer to merging before I think about documentation. Things are too open to change until then and I wouldn't want to waste your effort. I'll probably end up writing the documentation as well. I appreciate your offer!

**Comment by voltron4lyfe** [comment]
`--names: One or more instance names to sync.`
^^^ If not specified, does this sync all instances?  If so, then should add the same callout that all instances will be synced the same way it's called out for services. 

Small concern that `--names` may end up being ambiguous.  Thinking down the line there may be other things with names (e.g. quality profiles, release profiles, ????).  Maybe this should be `--instances`? Really being nit-picky though.  

Overall makes sense.  I can see how a future service (e.g. Lidarr, Readarr, ????) could slide in here, even if they aren't currently on the roadmap.

**Comment by rcdailey** [comment]
So in my CLI proposal, I suggested arranging options next to the appropriate segments/verbs of the command chain. For example:
```
recyclarr --debug sync radarr
```
As opposed to:
```
recyclarr sync radarr --debug
```
Is this format unusual for linux apps? I recall seeing this before, for example with `git -c`. I'm having a really hard time finding a CLI library that parses arguments like this. They all want you to push options to the right so that they are *after* all subcommands in the chain.

**Comment by yammes08** [comment]
I'm not massively familiar with it, beyond running commands as and when I need to (and usually following a guide or documentation). I found this though: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_01 - which you may well have read already. It seems to suggest flags after the application name, but having operands right at the end?

**Comment by rcdailey** [comment]
Resolved in cd6636d
